### PR TITLE
Add image upload support for rich-text editors (front-end part) #5915

### LINKED ISSUE #5915: Add image upload support for rich-text editors
Rich-text support was added throughout the web-interface, but image upload support wasn't implemented. So, it would be great to implement a solution that can be re-used in all the places with rich-text support.

### PR DESCRIPTION
**Pull request by belyabl9**
Fixes #5915

### CODE REVIEW

**Comment by belyabl9** [comment]
Ready to review

**Comment by damithc** [comment]
This has been pending review for 15 days! @wkurniawan07 
@belyabl9 you should have pinged to remind the reviewer.

**Comment by belyabl9** [comment]
@damithc , this issue will have a lot of rebase changes after https://github.com/TEAMMATES/teammates/pull/5888, so I wanted it to be merged before, so pinged that one

**Comment by damithc** [comment]
@wkurniawan07 this one needs a review.

File: src/main/java/teammates/common/util/Const.java [hunk 1]
```diff
@@ -977,6 +982,7 @@ public static final String getjQueryUiFilePath(boolean isDevServer) {
         public static final String ADMIN_EMAIL_CREATE_GROUP_RECEIVER_LIST_UPLOAD_URL =
                 "/admin/adminEmailCreateGroupReceiverListUploadUrl";
         
+        public static final String PUBLIC_FILE_SERVE = "/public/publicImageServe";
```
Comment (wkurniawan07) [inline src/main/java/teammates/common/util/Const.java]: `PUBLIC_IMAGE_SERVE` is a better name, even though the constant below has `FILE` in place of `IMAGE`.

**Comment by damithc** [comment]
What's the status of this PR?

**Comment by belyabl9** [comment]
@damithc , I'm still waiting for https://github.com/TEAMMATES/teammates/pull/5888 to be reviewed. Then I'll proceed with it

**Comment by wkurniawan07** [comment]
I think it's about time we get this up and running again :p

**Comment by belyabl9** [comment]
@wkurniawan07 , yep, I'll try to update the PR this week

**Comment by belyabl9** [comment]
@wkurniawan07 , ready to review

**Comment by belyabl9** [comment]
@wkurniawan07 , I agree, incremental approach will be easier for both of us. I think I'll return to this task these weekends

**Comment by belyabl9** [comment]
@wkurniawan07 , I've separated changes into two parts:
1) [Back-end](https://github.com/TEAMMATES/teammates/pull/6258) 
2) Front-end (this one)

AdminEmailGroupReceiverListUploadAction and StudentProfilePictureUploadAction changes will be investigated as a next step after these parts are merged.

Ready to review

File: src/main/webapp/js/instructorFeedbacks.js [hunk 2]
```diff
@@ -351,3 +337,21 @@ function collapseIfPrivateSession() {
...
+                selector: '#instructions',
+                inline: true,
+                readonly: isEdit,
+                fixed_toolbar_container: '#richtext-toolbar-container'
+            }
+        });
+        richTextEditor.init();
+        /* eslint-enable camelcase */
+    }
+});
```
Comment (wkurniawan07) [inline src/main/webapp/js/instructorFeedbacks.js]: Can you move this back up? `$(document).ready(...)` are all moved up as part of one of the refactorings done previously.

File: src/main/webapp/js/richTextEditor.js [hunk 3]
```diff
@@ -1,6 +1,49 @@
...
+function RichTextEditor(params) {
+
+    // ******************************************************************************************************
+    // Private properties
+    // ******************************************************************************************************
+
+    var MANDATORY_INIT_PARAMS = ['selector'];
+    var initParams = {};
+
+    var MANDATORY_ADDITIONAL_PARAMS = ['uploadImageId', 'uploadImageInputName', 'createImageUploadUrl'];
```
Comment (wkurniawan07) [inline src/main/webapp/js/richTextEditor.js]: These "additional" params look like they are specifically targeted for image uploads. Perhaps a better name is "image upload" params?

File: src/main/webapp/js/richTextEditor.js [hunk 4]
```diff
@@ -39,15 +89,129 @@ var richTextEditorBuilder = {
...
+
+    var createImageUploadUrl = function() {
+        $.ajax({
+            type: 'POST',
+            url: additionalParams.createImageUploadUrl,
+            beforeSend: function() {
+                showUploadingGif();
+            },
+            error: function() {
+                setErrorMessage('URL request failured, please try again.');
```
Comment (wkurniawan07) [inline src/main/webapp/js/richTextEditor.js]: fail~~ur~~ed

File: src/main/webapp/js/richTextEditor.js [hunk 5]
```diff
@@ -39,15 +89,129 @@ var richTextEditorBuilder = {
...
+        if ($('#' + uploadImageId + 'Form').length) {
+            return;
+        }
+
+        $('body').append($('<div id="uploadFileBlock" style="display: none;"></div>'));
+        $('#uploadFileBlock').append(
+                '<form id="' + uploadImageId + 'Form" action="" method="POST" enctype="multipart/form-data">');
+        $('#' + uploadImageId + 'Form').append('<span id="' + uploadImageId + 'Input" />');
+        $('#' + uploadImageId + 'Input').append(
+                '<input type="file" id="' + uploadImageId + '" name="' + uploadImageInputName + '" />');
```
Comment (wkurniawan07) [inline src/main/webapp/js/richTextEditor.js]: Closing with `/>` is against our style guide. Either close it with `>` for void elements, or use proper closing tags `</elementname>` for non-void elements. Unless XHTML is required in this operation?

File: src/main/webapp/js/richTextEditor.js [hunk 6]
```diff
@@ -1,6 +1,49 @@
...
+    // Private functions
+    // ******************************************************************************************************
+
+    var prepareInitParams = function() {
+        checkMandatoryParams(MANDATORY_INIT_PARAMS, initParams);
+    };
+
+    var prepareAdditionalParams = function() {
+        if (!additionalParams.uploadImageId) {
+            additionalParams.uploadImageId = 'uploadImage';
```
Comment (wkurniawan07) [inline src/main/webapp/js/richTextEditor.js]: Will `var = var || 'defaultValue';` work?

File: src/main/webapp/js/richTextEditor.js (hunk 3 above)
Comment (belyabl9) [inline src/main/webapp/js/richTextEditor.js]: I believe it's ok for it to be more abstract. It allows minimal changes in future, without permanent adjusting of the array name to its content

**Comment by belyabl9** [comment]
@wkurniawan07 , ready for review. Changesets are dependent, so they have to be merged simultaneously

**Comment by damithc** [comment]
Regarding merging order, shouldn't it be back-end first and then front-end? Back-end doesn't depend on front-end but front-end depends on back-end. The rule is, each PR should advance the master from one releasable state into a releasable state.
If the front-end PR is passing CI without the back-end, does that mean we don't have tests for the front end?

**Comment by belyabl9** [comment]
@damithc , you are right, such order should be used - (back-end -> front-end).
What about tests, as I understand we don't have tests for uploading images via rich-text editor

**Comment by damithc** [comment]
@belyabl9 
> What about tests, as I understand we don't have tests for uploading images via rich-text editor

Is there any problem preventing such tests? If not, we should add them. If not in this PR, immediately after this PR.

**Comment by belyabl9** [comment]
@damithc , currently I don't see any problems, it's just needed to discuss what sample file we will upload and where to put it. Let's do it immediately after this PR.

@wkurniawan07 , requested changes have been done, so, please, have a look

**Comment by belyabl9** [comment]
ping

**Comment by damithc** [comment]
@wkurniawan07 for your review please

**Comment by wkurniawan07** [comment]
Apologies, this has been slipping up my mind! I will give a review by tomorrow.

File: src/main/webapp/WEB-INF/tags/admin/email/compose.tag [hunk 7]
```diff
@@ -56,18 +56,4 @@
...
-        <form id="adminEmailFileForm" action=""
-            method="POST" enctype="multipart/form-data">
-            <span id="adminEmailFileInput"> <input
-                type="file"
-                name="<%=Const.ParamsNames.ADMIN_EMAIL_IMAGE_TO_UPLOAD%>"
-                id="adminEmailFile">
-            </span>
-        </form>
-
-        <div id="documentBaseUrl"><%=Config.APP_URL%></div>
```
Comment (wkurniawan07) [inline src/main/webapp/WEB-INF/tags/admin/email/compose.tag]: See if this import in the tag file is still needed. Similarly for `Const`.

File: src/main/webapp/js/richTextEditor.js (hunk 3 above)
Comment (wkurniawan07) [inline src/main/webapp/js/richTextEditor.js]: I agree that there is some truth to that, but there is a limit on being "more abstract". Here, your "additionalParams" makes all image-related parameters compulsory. Is any part of the name "additionalParams" suggesting that? In fact, the name part "additional" itself already suggests that all its parameters are compulsory, i.e. additional (fittingly).

Which is why I suggested something like "imageUploadParams" earlier. Granted that it is not as generalize-able as "additionalParams", it serves its purpose more clearly.

File: src/main/webapp/js/adminEmail.js [hunk 8]
```diff
@@ -7,24 +7,18 @@ $(document).ready(function() {
     
     $('.navbar-fixed-top').css('zIndex', 0);
 
-    /* eslint-disable camelcase */ // The property names are determined by external library (tinymce)
-    richTextEditorBuilder.initEditor('textarea', {
-        document_base_url: $('#documentBaseUrl').text(),
-        file_picker_callback: function(callback, value, meta) {
```
Comment (wkurniawan07) [inline src/main/webapp/js/adminEmail.js]: Is this being removed the correct thing to do?

File: src/main/webapp/js/adminEmail.js [hunk 9]
```diff
@@ -111,72 +105,6 @@ function submitGroupReceiverListUploadFormAjax() {
     clearUploadGroupReceiverListInfo();
 }
 
-function createImageUploadUrl() {
-    
-    $.ajax({
-        type: 'POST',
-        url: '/admin/adminEmailCreateImageUploadUrl',
-        beforeSend: function() {
-            showUploadingGif();
```
Comment (wkurniawan07) [inline src/main/webapp/js/adminEmail.js]: See if this function, if defined somewhere in `adminEmail.js`, is still used.

**Comment by damithc** [comment]
@belyabl9 shall re resume this?

**Comment by damithc** [comment]
@belyabl9 reminder

**Comment by damithc** [comment]
closing due to inactivity. @belyabl9 please reopen if you wish to resume

**Pull request closed without merging**